google-cloud-pubsub/lib/google/cloud/pubsub/publisher.rb

```diff
@@ -0,0 +1,195 @@
+              publish_batch_async topic_name, batch
+              # Create a new batch.
+              batch = QueuedBatch.new msg, block
+            end
+            @queue[topic_name] = batch
+          end
+          nil
+        end
+
+        def start
+          @publish_task.execute
+        end
+
+        def stop
+          flush
+          @publish_task.shutdown
+        end
+
+        def started?
+          @publish_task.running?
+        end
+
+        def stopping?
+          @publish_task.shuttingdown?
+        end
+
+        def stopped?
+          @publish_task.shutdown?
+        end
+
+        def wait timeout = nil
+          @publish_task.wait_for_termination timeout
+        end
+
+        def flush
+          synchronize do
+            @queue.each do |topic_name, batch|
+              publish_batch_async topic_name, batch
+            end
+            @queue = {}
+          end
+        end
+
+        protected
+
+        def create_pubsub_message data, attributes
+          attributes ||= {}
+          if data.is_a?(::Hash) && attributes.empty?
+            attributes = data
+            data = nil
+          end
+
+          # Convert IO-ish objects to strings
+          if data.respond_to?(:read) && data.respond_to?(:rewind)
+            data.rewind
+            data = data.read
+          end
+          # Convert data to encoded byte array to match the protobuf defn
+          data = String(data).force_encoding("ASCII-8BIT")
+
+          # Convert attributes to strings to match the protobuf definition
+          attributes = Hash[attributes.map { |k, v| [String(k), String(v)] }]
+
+          Google::Pubsub::V1::PubsubMessage.new data: data,
+                                                attributes: attributes
+        end
+
+        def publish_batch topic_name, batch
+          grpc = service.publish_messages topic_name, batch.messages
+          batch.items.zip(Array(grpc.message_ids)) do |item, id|
+            item.msg.message_id = id
+            item.block.call PublishResult.from_grpc(item.msg) if item.block
+          end
+        rescue => e
+          batch.items.each do |item|
+            item.block.call PublishResult.from_error(item.msg, e) if item.block
+          end
+        end
+
+        def publish_batch_async topic_name, batch
+          Concurrent::Future.new(executor: @thread_pool) do
+            publish_batch topic_name, batch
+          end.execute
+        end
+
+        class QueuedBatch
+          attr_reader :messages, :blocks
+
+          def initialize msg, block
+            @messages = [msg]
+            @blocks = [block]
+          end
+
+          def add msg, block
+            @messages << msg
+            @blocks << block
+          end
+
+          def try_add msg, block, max_count: MAX_COUNT, max_size: MAX_SIZE
+            return false if total_message_count + 1 > max_count
+            return false if total_message_size + msg.to_proto.size >= max_size
```

Comment:
Oh, I see where you came up with the 10 MB limit now. You are actually _failing the add_ when the next message goes over, which (I saw earlier) is what triggers the new batch.

I still think it would be better to have a smaller limit, just in case the client and server do not exactly agree on what the size is (e.g. if `msg.to_proto.size` has a bug due to a character encoding or some such -- it will totally happen). However, I recognize now that you accounted for the case I posited earlier.